File: backend/src/ICloudStore.Infrastructure/Services/SettingsService.cs
using ICloudStore.Application.Common;
using ICloudStore.Application.DTOs;
using ICloudStore.Application.Interfaces;
using ICloudStore.Domain.Entities;
using ICloudStore.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;

namespace ICloudStore.Infrastructure.Services;

public class SettingsService : ISettingsService
{
    private readonly ApplicationDbContext _context;

    public SettingsService(ApplicationDbContext context)
    {
        _context = context;
    }

    public async Task<ApiResponse<ThemeSettingDto>> GetThemeSettingAsync()
    {
        var theme = await _context.ThemeSettings.FirstOrDefaultAsync();
        if (theme == null)
        {
            theme = new ThemeSetting { ActiveTheme = "Light", AccentColor = "#25D366" };
            _context.ThemeSettings.Add(theme);
            await _context.SaveChangesAsync();
        }

        return ApiResponse<ThemeSettingDto>.SuccessResult(new ThemeSettingDto(
            theme.Id, theme.ActiveTheme, theme.AccentColor,
            theme.LogoLightPath, theme.LogoDarkPath, theme.FaviconPath));
    }

    public async Task<ApiResponse<ThemeSettingDto>> UpdateThemeSettingAsync(UpdateThemeSettingDto dto)
    {
        var theme = await _context.ThemeSettings.FirstOrDefaultAsync();
        if (theme == null)
        {
            theme = new ThemeSetting();
            _context.ThemeSettings.Add(theme);
        }

        theme.ActiveTheme = dto.ActiveTheme;
        theme.AccentColor = dto.AccentColor;
        await _context.SaveChangesAsync();

        return ApiResponse<ThemeSettingDto>.SuccessResult(new ThemeSettingDto(
            theme.Id, theme.ActiveTheme, theme.AccentColor,
            theme.LogoLightPath, theme.LogoDarkPath, theme.FaviconPath));
    }

    public async Task<ApiResponse<bool>> UpdateLogoAsync(string? logoLightPath, string? logoDarkPath, string? faviconPath)
    {
        var theme = await _context.ThemeSettings.FirstOrDefaultAsync();
        if (theme == null)
        {
            theme = new ThemeSetting();
            _context.ThemeSettings.Add(theme);
        }

        if (logoLightPath != null) theme.LogoLightPath = logoLightPath;
        if (logoDarkPath != null) theme.LogoDarkPath = logoDarkPath;
        if (faviconPath != null) theme.FaviconPath = faviconPath;

        await _context.SaveChangesAsync();
        return ApiResponse<bool>.SuccessResult(true);
    }

    public async Task<ApiResponse<List<StoreSettingDto>>> GetAllStoreSettingsAsync()
    {
        var settings = await _context.StoreSettings.ToListAsync();
        return ApiResponse<List<StoreSettingDto>>.SuccessResult(
            settings.Select(s => new StoreSettingDto(s.Id, s.Key, s.ValueAr, s.ValueEn)).ToList());
    }

    public async Task<ApiResponse<StoreSettingDto>> GetStoreSettingAsync(string key)
    {
        var setting = await _context.StoreSettings.FirstOrDefaultAsync(s => s.Key == key);
        if (setting == null)
            return ApiResponse<StoreSettingDto>.FailResult("الإعداد غير موجود");

        return ApiResponse<StoreSettingDto>.SuccessResult(
            new StoreSettingDto(setting.Id, setting.Key, setting.ValueAr, setting.ValueEn));
    }

    public async Task<ApiResponse<StoreSettingDto>> UpdateStoreSettingAsync(string key, UpdateStoreSettingDto dto)
    {
        var setting = await _context.StoreSettings.FirstOrDefaultAsync(s => s.Key == key);
        if (setting == null)
        {
            setting = new StoreSetting { Key = key };
            _context.StoreSettings.Add(setting);
        }

        setting.ValueAr = dto.ValueAr;
        setting.ValueEn = dto.ValueEn;
        await _context.SaveChangesAsync();

        return ApiResponse<StoreSettingDto>.SuccessResult(
            new StoreSettingDto(setting.Id, setting.Key, setting.ValueAr, setting.ValueEn));
    }

    public async Task<ApiResponse<List<SocialLinkDto>>> GetAllSocialLinksAsync(bool includeHidden = false)
    {
        var query = _context.SocialLinks.AsQueryable();
        if (!includeHidden)
            query = query.Where(s => s.IsVisible);

        var links = await query.OrderBy(s => s.SortOrder).ToListAsync();
        return ApiResponse<List<SocialLinkDto>>.SuccessResult(
            links.Select(s => new SocialLinkDto(s.Id, s.Platform, s.Url, s.IconClass, s.IsVisible, s.SortOrder)).ToList());
    }

    public async Task<ApiResponse<SocialLinkDto>> CreateSocialLinkAsync(CreateSocialLinkDto dto)
    {
        var link = new SocialLink
        {
            Platform = dto.Platform,
            Url = dto.Url,
            IconClass = dto.IconClass,
            IsVisible = dto.IsVisible,
            SortOrder = dto.SortOrder
        };

        _context.SocialLinks.Add(link);
        await _context.SaveChangesAsync();

        return ApiResponse<SocialLinkDto>.SuccessResult(
            new SocialLinkDto(link.Id, link.Platform, link.Url, link.IconClass, link.IsVisible, link.SortOrder));
    }

    public async Task<ApiResponse<SocialLinkDto>> UpdateSocialLinkAsync(int id, UpdateSocialLinkDto dto)
    {
        var link = await _context.SocialLinks.FindAsync(id);
        if (link == null)
            return ApiResponse<SocialLinkDto>.FailResult("الرابط غير موجود");

        link.Platform = dto.Platform;
        link.Url = dto.Url;
        link.IconClass = dto.IconClass;
        link.IsVisible = dto.IsVisible;
        link.SortOrder = dto.SortOrder;
        await _context.SaveChangesAsync();

        return ApiResponse<SocialLinkDto>.SuccessResult(
            new SocialLinkDto(link.Id, link.Platform, link.Url, link.IconClass, link.IsVisible, link.SortOrder));
    }

    public async Task<ApiResponse<bool>> DeleteSocialLinkAsync(int id)
    {
        var link = await _context.SocialLinks.FindAsync(id);
        if (link == null)
            return ApiResponse<bool>.FailResult("الرابط غير موجود");

        _context.SocialLinks.Remove(link);
        await _context.SaveChangesAsync();

        return ApiResponse<bool>.SuccessResult(true);
    }

    public async Task<ApiResponse<bool>> ReorderSocialLinksAsync(List<int> linkIds)
    {
        for (int i = 0; i < linkIds.Count; i++)
        {
            var link = await _context.SocialLinks.FindAsync(linkIds[i]);
            if (link != null) link.SortOrder = i + 1;
        }
        await _context.SaveChangesAsync();
        return ApiResponse<bool>.SuccessResult(true);
    }

    public async Task<ApiResponse<StoreInfoDto>> GetPublicStoreInfoAsync()
    {
        var settings = await _context.StoreSettings.ToListAsync();
        var socialLinks = await _context.SocialLinks.Where(s => s.IsVisible).OrderBy(s => s.SortOrder).ToListAsync();
        var theme = await _context.ThemeSettings.FirstOrDefaultAsync();

        var getValue = (string key) => settings.FirstOrDefault(s => s.Key == key);

        return ApiResponse<StoreInfoDto>.SuccessResult(new StoreInfoDto(
            getValue("StoreName")?.ValueAr,
            getValue("StoreName")?.ValueEn,
            getValue("WhatsAppNumber")?.ValueAr,
            getValue("PhoneNumber")?.ValueAr,
            getValue("SupportEmail")?.ValueAr,
            getValue("Address")?.ValueAr,
            getValue("Address")?.ValueEn,
            getValue("WorkingHours")?.ValueAr,
            getValue("WorkingHours")?.ValueEn,
            socialLinks.Select(s => new SocialLinkDto(s.Id, s.Platform, s.Url, s.IconClass, s.IsVisible, s.SortOrder)).ToList(),
            theme != null
                ? new ThemeSettingDto(theme.Id, theme.ActiveTheme, theme.AccentColor, theme.LogoLightPath, theme.LogoDarkPath, theme.FaviconPath)
                : new ThemeSettingDto(0, "Light", "#25D366", null, null, null)
        ));
    }
}

File: frontend/admin/src/app/core/services/settings.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, forkJoin, map, of, catchError } from 'rxjs';
import { environment } from '@environments/environment';

// Local simplified ApiResponse (settings-specific)
interface SettingsApiResponse<T> {
  success: boolean;
  data?: T;
  message?: string;
  errors?: string[];
}

// Matches backend StoreSettingDto
export interface StoreSetting {
  id: number;
  key: string;
  valueAr?: string;
  valueEn?: string;
}

// Matches backend ThemeSettingDto
export interface ThemeSetting {
  id: number;
  activeTheme: string;
  accentColor: string;
  logoLightPath?: string;
  logoDarkPath?: string;
  faviconPath?: string;
}

// Matches backend SocialLinkDto
export interface SocialLink {
  id: number;
  platform: string;
  url: string;
  iconClass?: string;
  isVisible: boolean;
  sortOrder: number;
}

// Combined settings for the admin UI
export interface StoreSettings {
  // Store Info (from key-value settings)
  storeNameAr?: string;
  storeNameEn?: string;
  taglineAr?: string;
  taglineEn?: string;
  descriptionAr?: string;
  descriptionEn?: string;
  
  // Contact
  email?: string;
  phone?: string;
  whatsAppNumber?: string;
  addressAr?: string;
  addressEn?: string;
  
  // Social Links
  facebookUrl?: string;
  instagramUrl?: string;
  twitterUrl?: string;
  youtubeUrl?: string;
  tiktokUrl?: string;
  
  // Theme
  primaryColor?: string;
  secondaryColor?: string;
  logoUrl?: string;
  faviconUrl?: string;
  
  // SEO
  metaTitleAr?: string;
  metaTitleEn?: string;
  metaDescriptionAr?: string;
  metaDescriptionEn?: string;
  metaKeywordsAr?: string;
  metaKeywordsEn?: string;
  
  // Business
  workingHoursAr?: string;
  workingHoursEn?: string;
  currency?: string;
  defaultLanguage?: string;
}

export type UpdateSettingsRequest = Partial<StoreSettings>;

@Injectable({
  providedIn: 'root'
})
export class SettingsService {
  private readonly apiUrl = `${environment.apiUrl}/Settings`;

  constructor(private http: HttpClient) {}

  // Get combined settings from multiple endpoints
  get(): Observable<SettingsApiResponse<StoreSettings>> {
    return forkJoin({
      storeSettings: this.http.get<SettingsApiResponse<StoreSetting[]>>(`${this.apiUrl}/store`).pipe(
        catchError(() => of({ success: true, data: [] as StoreSetting[] }))
      ),
      theme: this.http.get<SettingsApiResponse<ThemeSetting>>(`${this.apiUrl}/theme`).pipe(
        catchError(() => of({ success: true, data: undefined }))
      ),
      socialLinks: this.http.get<SettingsApiResponse<SocialLink[]>>(`${this.apiUrl}/social-links`).pipe(
        catchError(() => of({ success: true, data: [] as SocialLink[] }))
      )
    }).pipe(
      map(results => {
        const settings: StoreSettings = {};
        
        // Map store settings from key-value pairs
        if (results.storeSettings.data) {
          for (const setting of results.storeSettings.data) {
            switch (setting.key) {
              case 'store_name': 
                settings.storeNameAr = setting.valueAr; 
                settings.storeNameEn = setting.valueEn; 
                break;
              case 'tagline': 
                settings.taglineAr = setting.valueAr; 
                settings.taglineEn = setting.valueEn; 
                break;
              case 'description':
                settings.descriptionAr = setting.valueAr;
                settings.descriptionEn = setting.valueEn;
                break;
              case 'email': settings.email = setting.valueEn || setting.valueAr; break;
              case 'phone': settings.phone = setting.valueEn || setting.valueAr; break;
              case 'whatsapp': settings.whatsAppNumber = setting.valueEn || setting.valueAr; break;
              case 'address':
                settings.addressAr = setting.valueAr;
                settings.addressEn = setting.valueEn;
                break;
              case 'working_hours':
                settings.workingHoursAr = setting.valueAr;
                settings.workingHoursEn = setting.valueEn;
                break;
              case 'currency': settings.currency = setting.valueEn || 'EGP'; break;
              case 'default_language': settings.defaultLanguage = setting.valueEn || 'ar'; break;
              case 'meta_title':
                settings.metaTitleAr = setting.valueAr;
                settings.metaTitleEn = setting.valueEn;
                break;
              case 'meta_description':
                settings.metaDescriptionAr = setting.valueAr;
                settings.metaDescriptionEn = setting.valueEn;
                break;
              case 'meta_keywords':
                settings.metaKeywordsAr = setting.valueAr;
                settings.metaKeywordsEn = setting.valueEn;
                break;
            }
          }
        }
        
        // Map theme settings
        if (results.theme.data) {
          settings.primaryColor = results.theme.data.accentColor || '#25D366';
          settings.logoUrl = results.theme.data.logoLightPath;
          settings.faviconUrl = results.theme.data.faviconPath;
        }
        
        // Map social links
        if (results.socialLinks.data) {
          for (const link of results.socialLinks.data) {
            switch (link.platform.toLowerCase()) {
              case 'facebook': settings.facebookUrl = link.url; break;
              case 'instagram': settings.instagramUrl = link.url; break;
              case 'twitter': settings.twitterUrl = link.url; break;
              case 'youtube': settings.youtubeUrl = link.url; break;
              case 'tiktok': settings.tiktokUrl = link.url; break;
            }
          }
        }
        
        return { success: true, data: settings };
      })
    );
  }

  // Update settings by calling multiple endpoints
  update(settings: UpdateSettingsRequest): Observable<SettingsApiResponse<StoreSettings>> {
    const updates: Observable<any>[] = [];
    
    // Update store settings as key-value pairs
    const storeSettings: { key: string; valueAr?: string; valueEn?: string }[] = [];
    
    if (settings.storeNameAr !== undefined || settings.storeNameEn !== undefined) {
      storeSettings.push({ key: 'store_name', valueAr: settings.storeNameAr, valueEn: settings.storeNameEn });
    }
    if (settings.taglineAr !== undefined || settings.taglineEn !== undefined) {
      storeSettings.push({ key: 'tagline', valueAr: settings.taglineAr, valueEn: settings.taglineEn });
    }
    if (settings.email !== undefined) {
      storeSettings.push({ key: 'email', valueEn: settings.email });
    }
    if (settings.phone !== undefined) {
      storeSettings.push({ key: 'phone', valueEn: settings.phone });
    }
    if (settings.whatsAppNumber !== undefined) {
      storeSettings.push({ key: 'whatsapp', valueEn: settings.whatsAppNumber });
    }
    if (settings.addressAr !== undefined || settings.addressEn !== undefined) {
      storeSettings.push({ key: 'address', valueAr: settings.addressAr, valueEn: settings.addressEn });
    }
    if (settings.workingHoursAr !== undefined || settings.workingHoursEn !== undefined) {
      storeSettings.push({ key: 'working_hours', valueAr: settings.workingHoursAr, valueEn: settings.workingHoursEn });
    }
    if (settings.currency !== undefined) {
      storeSettings.push({ key: 'currency', valueEn: settings.currency });
    }
    if (settings.defaultLanguage !== undefined) {
      storeSettings.push({ key: 'default_language', valueEn: settings.defaultLanguage });
    }
    
    // Add store setting updates
    for (const s of storeSettings) {
      updates.push(
        this.http.put(`${this.apiUrl}/store/${s.key}`, { valueAr: s.valueAr, valueEn: s.valueEn }).pipe(
          catchError(() => of({ success: false }))
        )
      );
    }
    
    // Update theme if color changed
    if (settings.primaryColor !== undefined) {
      updates.push(
        this.http.put(`${this.apiUrl}/theme`, { 
          activeTheme: 'Light', 
          accentColor: settings.primaryColor 
        }).pipe(catchError(() => of({ success: false })))
      );
    }
    
    if (updates.length === 0) {
      return of({ success: true, data: settings as StoreSettings, message: 'لا توجد تغييرات' });
    }
    
    return forkJoin(updates).pipe(
      map(() => ({ success: true, data: settings as StoreSettings, message: 'تم حفظ الإعدادات' }))
    );
  }

  // Get theme settings only
  getTheme(): Observable<SettingsApiResponse<ThemeSetting>> {
    return this.http.get<SettingsApiResponse<ThemeSetting>>(`${this.apiUrl}/theme`);
  }

  // Update theme
  updateTheme(theme: { activeTheme: string; accentColor: string }): Observable<SettingsApiResponse<ThemeSetting>> {
    return this.http.put<SettingsApiResponse<ThemeSetting>>(`${this.apiUrl}/theme`, theme);
  }

  // Get social links
  getSocialLinks(): Observable<SettingsApiResponse<SocialLink[]>> {
    return this.http.get<SettingsApiResponse<SocialLink[]>>(`${this.apiUrl}/social-links`);
  }
}
